### ISSUE TITLE
[rllib] custom policy networks (custom weights initializers)

### ISSUE THREAD

**Issue by Scitator** [ray-project/ray#2423]
Hi,

For some situations, for example when we use selu activation functions, we need to use custom initializers instead of "typical" xavier.

For observation processing models all look quite simple - we can just define our custom model, but when it comes to policy model ([example](https://github.com/ray-project/ray/blob/master/python/ray/rllib/agents/ddpg/ddpg_policy_graph.py#L25))....it's a bit more complicated.

So, it looks like good feature to give an opportunity to configure weight initialisation, nevertheless, it's a bit tricky. For now, I see 3 ways to do this:
1. just specify something like `initializers.xavier_initializer(uniform=True, seed=42)` in config and eval it at network creation (hacker way).
2. specify something like:
```
"initializer": {
    "initializer": xavier_initializer,
    "uniform": true,
    "seed": 42
}
```
I use such configuration in my projects.
3. add possibility to use custom policy networks? or just q/value networks?

If you want, I can PR for more Ray-way solution.
For now, I see several files of interest:
- https://github.com/ray-project/ray/blob/master/python/ray/rllib/agents/ddpg/ddpg_policy_graph.py
- https://github.com/ray-project/ray/blob/master/python/ray/rllib/models/fcnet.py
- https://github.com/ray-project/ray/blob/master/python/ray/rllib/models/visionnet.py
- and other rllib agents.

PS. I wonder, why do you use `normc_initializer` for `FullyConnectedNetwork`? Is it works better than xavier?
PS2. Why do you use `tf.contrib.layers` instead of `tf.layers`? Are they better of it is just legacy code?

**Comment by Scitator** [comment]
I am not sure if it the correct way, but..... all works great with simple copying `ddpg_policy_graph.py` and modifying it as necessary, then:
```
from ddpg.ddpg_policy_graph import DDPGPolicyGraph as PolicyGraph
from ray.rllib.agents.ddpg.apex import ApexDDPGAgent as Agent
from ray.rllib.agents.ddpg.apex import APEX_DDPG_DEFAULT_CONFIG as CONFIG

Agent._policy_graph = PolicyGraph
```
works quite nice.

It definitely not the perfect solution, but due to unstandartiastion of DNQ policies graph it looks like quite good working hack :). After looking the source code, it looks like it's better to firstly merge DQN's policy graphs to something standard and then add possibility for custom q/value networks.

**Comment by ericl** [comment]
Oh, that's a good solution (if a bit hacky indeed).

I'm not sure what you mean though about DDPG not having a standard graph,
it does extend TFPolicyGraph.

**Comment by Scitator** [comment]
@ericl I just propose to add something like `_init` to `PolicyGraph `, so you can override main used nets (like q/value networks) without changing the whole graph.
For this we need to standardize networks input/output (just like with custom models).

another solutions is to register custom q/value networks in Ray and use them if `custom_q_networks` flag is not None. 

What's the most Ray-way?

**Comment by ericl** [comment]
Hm, what it we made the _build_p_network functions etc methods of the policy graph class, so they can be easily overridden by subclasses? Then, we can document for DQN/DDPG that this can be done to provide custom internal networks (or you can swap the whole policy graph class for deeper customization).

I am also fine with making them config flags for DQN/DDPG.

**Comment by Scitator** [comment]
merged
